docs/src/main/paradox/developer/eventing/event-acknowledgments.md

```diff
@@ -0,0 +1,48 @@
+
+We will assume a fairly well understood problem domain to frame our use cases in. We have an event sourced entity that represents a user, which is used to provide a user management service. This user entity might have events such as `UserCreated`, `PasswordChanged`, `EmailChanged`. These events are how the user entity internally represents a user, and their schema should be able to change freely to accommodate the needs of the user entity without affecting any downstream services that might consume these events. Migrations may be defined on them to evolve these events over time. For this reason, this stream is not published directly to external services, rather, an anti-corruption layer is used, that translates these events to an external representation of the same events.
+
+Unlike the internal events from the event log, the external events cannot be changed freely, their schema forms a contract with the downstream services that consumes them. When the internal events change in their structure, the external events must not change in a way that will break backwards compatibility.
+
+Note that in the use cases below, we're not trying to build a coherent model of a user management service, nor are we trying to design the perfect user management service. Rather, we're saying "let's assume that the business has requirement X", and then discuss how we might provide the mechanism to implement that. These requirements are a given, and hopefully are intuitively relatable, even if that's not how you would necessarily design a system to function. The point is, it's up to the business to make its own decisions, and Cloudstate to provide the mechanisms to make those decisions possible to implement, whatever they are.
+
+#### Filtering
+
+Not all events in the internal model will necessarily map to events in the external model. For example, a business decision might be that a users password might be the domain of the user management service and the user management service alone. When a `PasswordChanged` event is emitted from the event log, no event should then be published externally. Effectively, the `PasswordChanged` event must be filtered from the event stream.
+
+The challenge in acknowledgement here is when should the `PasswordChanged` event be acknowledged, and how should the user function indicate this? Since no event is being output, the output of an event cannot be used to indicate that the `PasswordChanged` event should be acknowledged. Rather, it's the absence of an output event that should indicate that a `PasswordChanged` event should be acknowledged. But how do we differentiate between on output event, and slow processing of the input event? Does the Cloudstate proxy say "well, you haven't output anything for 10 seconds, so I'm going to acknowledge?". This puts a 10 second latency on the processing of all `PasswordChanged` events, so that solution doesn't work. Do we have a specific ack signal, which accompanies no event?
```

Comment:
btw... I initially wrote this document just to collect my own thoughts. I didn't finish it, and it's just a raw collection of thoughts. When I created this PR, I noticed the document was there, and thought well maybe it'll be useful to help others understand the context.

So, the solution I ended up going with was the following:

* If all you want to do is emit 0-1 events, then you can define unary gRPC calls, and respond with a reply if you want to emit one event, or with no reply for zero. The ack signal is the same in either case - the response, either containing a reply or not.
* If you want to emit 0-many events, then you can define a server-side streamed gRPC call, where the event is sent as the single input, and you can send 0 to many replies in gRPC response, followed by closing the stream. The close of the stream signals the acknowledgement.
* If you define a client streamed gRPC call (either one direction or duplex), it will treat it the same as a server side streamed gRPC call, it will only send one event into the client stream, and consider termination of the output stream, or redeem of the output unary value, as the acknowledge event.
* Finally, you can also emit zero to many events from unary calls using effects that forward to other commands with an output handler, though right now, if they fail to be sent, that won't cause a failure on the stream so the event will be acknowledged. We need to decide whether synchronous events should impact the result if they fail (currently they don't, the processor does wait for them to complete, but ignores the success/failure result), or if we need a new semantic to indicate "do this effect, and if it fails, fail the whole command".

Reply:
@jroper I agree with this line of thinking. I personally believe that explicit acking is going to be a recipe for a lot of bugs, so steering clear of that seems like a big win.